torch/testing/_internal/optests/generate_tests.py

```diff
@@ -0,0 +1,318 @@
+                f"has method named {new_method_name}. Please rename the original "
+                f"method on the TestCase."
+            )
+        setattr(testcase, new_method_name, new_method)
+
+    test_utils = {name: ALL_TEST_UTILS[name] for name in test_utils}
+    for attr in test_methods:
+        for prefix, tester in test_utils.items():
+            construct_method(attr, prefix, tester)
+
+
+TEST_OPTIONS = ("xfail", "skip", "success")
+
+
+def validate_failures_dict(failure_dict, test_utils, testcase):
+    """Validates the failures dict.
+
+    The failure dict looks something like the following.
+    It maps operator name (qualname) to a list of autogenerated tests.
+    Each autogenerated test may have a check for the operator (if the operator is
+    called by the test); the dictionary specifies if we should skip the check,
+    or if we expect some check to fail.
+
+    {
+        "fbgemm::split_lengths": {
+            "test_schema__test_split_lengths": "xfail",
+            "test_schema__test_split_lengths_empty": "skip",
+        }
+        "fbgemm::gather_lengths": {
+            "test_schema__test_gather_lengths": "xfail",
+        }
+    }
+
+    We require that all keys are sorted in alphabetical order. This makes
+    it easier for us to codemod the failures_dict.
+    """
+    qualnames = list(failure_dict.keys())
+    if qualnames != sorted(qualnames):
+        raise RuntimeError("The failures dict must be sorted in alphabetical order")
+    for qualname, test_to_option in failure_dict.items():
+        test_names = list(test_to_option.keys())
+        if test_names != sorted(test_names):
+            raise RuntimeError(
+                f"failures_dict['{qualname}']'s keys must be sorted in alphabetical order"
+            )
+        for test_name, test_option in test_to_option.items():
+            if test_option not in TEST_OPTIONS:
+                raise RuntimeError(
+                    f"In failures_dict, got value={test_option} but it needs to be in {TEST_OPTIONS}"
+                )
+            if not any(test_name.startswith(test) for test in test_utils):
+                raise RuntimeError(
+                    f"In failures_dict, test name '{test_name}' should begin with one of {test_utils}"
+                )
+            for test in test_utils:
+                if not test_name.startswith(test):
+                    continue
+                base_test_name = test_name[len(test) + 2 :]
+                if hasattr(testcase, base_test_name):
+                    continue
+                raise RuntimeError(
+                    f"In failures dict, got test name '{test_name}'. We parsed this as "
+                    f"running test '{test}' on '{base_test_name}', but "
+                    f"{base_test_name} does not exist on the TestCase. "
+                    f"Maybe you need to change the test name?"
+                )
+
+
+class OpCheckMode(TorchFunctionMode):
```

Comment:
On consequence of having every test be a separate torch function mode is if your test originally took T time, now it takes T * K time, where K is the number of function modes we need to iterate through. It doesn't need to take this long though: all you're really doing is using the pre-existing test code to get real input values to run the operators with, so if memory usage were no concern you could hypothetically just run through the tests once, accurately cloning all of the inputs, and then use these inputs to run through the rest of the tests. You could even use this to do some deduplication (since none of your torch function modes are data dependent, you can keep only one version for any given size that matches).

Granted, this is somewhat difficult to engineer because it's not very natural to have a test suite that runs a bunch of tests first, and THEN runs the rest of the tests. But I feel like there may be ways around this? For example, if you're OK with only running all the tests twice (still an improvement), you can run the tests normally, and then run the tests to setup a fixture, representing all of the test inputs. Then each successive test case picks out what particular input they want to test. Or you could even just stack all of the torch function modes at the same time.

I'm somewhat sensitive to this because when I was working with cross-ref tests I was running them on the PyTorch test suite, which takes a quite long time to run. Maybe all of the internal tests you are working with are quite quick to run; this is less of a concern in that case. (Deduplication was especially a problem with cross-ref tests, since ops were continuously reused to drive other tests). But it is not hard to imagine someone writing a fairly comprehensive test suite and this mechanism quintupling the runtime of those tests. This is not really a good incentive.

Reply:
> On consequence of having every test be a separate torch function mode is if your test originally took T time, now it takes T * K time, where K is the number of function modes we need to iterate through.

Agree

>  and then run the tests to setup a fixture, representing all of the test inputs. Then each successive test case picks out what particular input they want to test. 

I thought about doing this, but I had trouble engineering this:
- If we run all of the tests a second time before auto-generating tests, then that means it takes the length of the test suite run (20-40 minutes in some cases) before you are actually able to run the autogenerated tests.
- During the second run, we could run the tests one at a time, e.g. test_mm, and then generate additional tests from it. This seems reasonable (and doable in OSS), but it is a bit difficult to dynamically generate tests internally

> Or you could even just stack all of the torch function modes at the same time.

If I am understanding this correctly, your suggestion is when we run `test_mm` the second time, we run all of our tests (schema_check, aot_autograd, ...) during that run. This means that the second `test_mm` run takes a long time and there is no mechanism to separate out what tests we're running (unless we build something to control that).